Mathlib/SetTheory/Ordinal/Veblen.lean

```diff
@@ -0,0 +1,317 @@
+  (veblenWith_right_strictMono hf o).injective
+
+theorem veblenWith_inj : veblenWith f o a = veblenWith f o b ↔ a = b :=
+  (veblenWith_injective hf o).eq_iff
+
+theorem right_le_veblenWith (a b : Ordinal) : b ≤ veblenWith f a b :=
+  (veblenWith_right_strictMono hf a).le_apply
+
+theorem veblenWith_left_monotone (o : Ordinal) : Monotone (veblenWith f · o) := by
+  rw [monotone_iff_forall_lt]
+  intro a b h
+  rw [← veblenWith_veblenWith_of_lt hf h]
+  exact (veblenWith_right_strictMono hf a).monotone (right_le_veblenWith hf b o)
+
+theorem veblenWith_pos (hp : 0 < f 0) (a b : Ordinal) : 0 < veblenWith f a b := by
+  have H (b) : 0 < veblenWith f 0 b := by
+    rw [veblenWith_zero]
+    exact hp.trans_le (hf.monotone (Ordinal.zero_le _))
+  obtain rfl | h := Ordinal.eq_zero_or_pos a
+  · exact H b
+  · rw [← veblenWith_veblenWith_of_lt hf h]
+    exact H _
+
+theorem veblenWith_zero_strictMono (hp : 0 < f 0) : StrictMono (veblenWith f · 0) := by
+  intro a b h
+  dsimp only
+  rw [← veblenWith_veblenWith_of_lt hf h, veblenWith_lt_veblenWith_iff_right hf]
+  exact veblenWith_pos hf hp b 0
+
+theorem veblenWith_zero_lt_veblenWith_zero_iff_left (hp : 0 < f 0) :
+    veblenWith f a 0 < veblenWith f b 0 ↔ a < b :=
+  (veblenWith_zero_strictMono hf hp).lt_iff_lt
+
+theorem veblenWith_zero_le_veblenWith_zero_iff_left (hp : 0 < f 0) :
+    veblenWith f a 0 ≤ veblenWith f b 0 ↔ a ≤ b :=
+  (veblenWith_zero_strictMono hf hp).le_iff_le
+
+theorem veblenWith_zero_inj (hp : 0 < f 0) : veblenWith f a 0 = veblenWith f b 0 ↔ a = b :=
+  (veblenWith_zero_strictMono hf hp).injective.eq_iff
+
+theorem left_le_veblenWith (hp : 0 < f 0) (a b : Ordinal) : a ≤ veblenWith f a b :=
+  (veblenWith_zero_strictMono hf hp).le_apply.trans <|
+    (veblenWith_right_strictMono hf _).monotone (Ordinal.zero_le _)
+
+theorem IsNormal.veblenWith_zero (hp : 0 < f 0) : IsNormal (veblenWith f · 0) := by
+  rw [isNormal_iff_strictMono_limit]
+  refine ⟨veblenWith_zero_strictMono hf hp, fun o ho a IH ↦ ?_⟩
+  rw [veblenWith_of_ne_zero f ho.pos.ne', derivFamily_zero]
+  apply nfpFamily_le fun l ↦ ?_
+  suffices ∃ b < o, List.foldr _ 0 l ≤ veblenWith f b 0 by
+    obtain ⟨b, hb, hb'⟩ := this
+    exact hb'.trans (IH b hb)
+  induction l with
+  | nil => use 0; simp [ho.pos]
+  | cons a l IH =>
+    obtain ⟨b, hb, hb'⟩ := IH
+    refine ⟨_, ho.succ_lt (max_lt a.2 hb), ((veblenWith_right_strictMono hf _).monotone <|
+      hb'.trans <| veblenWith_left_monotone hf _ <|
+        (le_max_right a.1 b).trans (Order.le_succ _)).trans ?_⟩
+    rw [veblenWith_veblenWith_of_lt hf]
+    rw [Order.lt_succ_iff]
+    exact le_max_left _ b
+
+/-- `veblenWith f a b < veblenWith f c d` iff one of the following holds:
+* `a = c` and `b < d`
+* `a < c` and `b < veblenWith f c d`
+* `a > c` and `veblenWith f a b < d` -/
+theorem veblenWith_lt_veblenWith_iff : veblenWith f a b < veblenWith f c d ↔
+    a = c ∧ b < d ∨ a < c ∧ b < veblenWith f c d ∨ c < a ∧ veblenWith f a b < d := by
```

Comment:
Please start the type of a theorem on a new line if it's going to need a line break
```suggestion
theorem veblenWith_lt_veblenWith_iff :
    veblenWith f a b < veblenWith f c d ↔
      a = c ∧ b < d ∨ a < c ∧ b < veblenWith f c d ∨ c < a ∧ veblenWith f a b < d := by
```